File: Plutus/Frontend/Plutus.Frontend.WebApp/src/reporting/publishedReports.ts
/**
 * Which reports the portal has published to this till — ruling 5b(a), 2026-08-19.
 *
 * ⚠⚠ Matt: *"Portal shows which reports a till can show."* The catalogue is now the SUPERSET; this
 * decides which of it an operator is offered.
 *
 * ⚠⚠ C2 TWIN of `SharedKernel/ReportCatalogue.PublishedOr` + MAUI's `Services/Reporting/
 * PublishedReports`. All three answer the same two questions — what does silence mean, and what
 * happens to a key we do not recognise — and two tills that disagree show different menus for the same
 * shop. The vectors run on both sides: `publishedReports.test.ts` here, `ReportCatalogueTests` in C#.
 *
 * ⚠⚠ **THE PUBLISH DECIDES THE MENU, THE PERMISSION DECIDES THE DOOR.** This module knows nothing about
 * permissions and must not learn: `ReportingPage` applies both, and a report published but unreadable is
 * not listed at all — never listed-and-refused, because a greyed tab leaks what other roles can see.
 */

/**
 * The full catalogue, in menu order.
 *
 * ⚠⚠ MUST MATCH `SharedKernel.ReportCatalogue.Keys` — order included, because it IS the menu order on
 * both tills. The C# side pins its list in a test precisely because these keys are stored in the
 * database; renaming one silently unpublishes it for every tenant that had it ticked.
 */
export const ALL_REPORT_KEYS = [
  "summary", "vat", "items-sold", "category-sales",
  "best-sellers", "stock", "negative-stock", "sales",
] as const;

/** Versioned, so a future change of shape cannot be read as a valid old value. */
const CACHE_KEY = "plutus.reports.published.v1";

/**
 * What the till may offer, given what the server said.
 *
 * ⚠⚠ NULL MEANS "NOBODY HAS CHOSEN" AND RESOLVES TO EVERY REPORT. A tenant who has never opened the
 * portal screen must see the menu they saw yesterday — defaulting to "nothing" would empty the Reports
 * tab in every shop the moment this deployed. **Absence of a choice is not a choice.**
 *
 * ⚠ An EMPTY array is a real decision — "this till shows no reports" — and is honoured. That is why the
 * never-chosen case is `null` and not `[]`: collapsing them would make one impossible to express.
 *
 * ⚠ Unknown keys are DROPPED, and the result is in CATALOGUE order rather than stored order, so the
 * menu cannot be reshuffled by however the portal happened to serialise the list.
 */
export function publishedOr(storedKeys: readonly string[] | null | undefined): string[] {
  if (storedKeys === null || storedKeys === undefined) return [...ALL_REPORT_KEYS];

  const stored = new Set(storedKeys);
  return ALL_REPORT_KEYS.filter((k) => stored.has(k));
}

/**
 * The last answer this browser had, or null if it has never had one.
 *
 * ⚠ A corrupt cache reads as "never asked", which shows every report. Failing towards MORE reports is
 * the right direction for a menu; failing towards none is a till that looks broken.
 */
export function cachedPublished(): string[] | null {
  try {
    const raw = localStorage.getItem(CACHE_KEY);
    if (raw === null) return null;
    const parsed = JSON.parse(raw);
    return Array.isArray(parsed) ? parsed.filter((k) => typeof k === "string") : null;
  } catch {
    return null;
  }
}

export function cachePublished(keys: readonly string[]): void {
  try {
    localStorage.setItem(CACHE_KEY, JSON.stringify(keys));
  } catch {
    // ⚠ A full or blocked localStorage must not stop the Reports tab rendering.
  }
}

/**
 * Ask the server, cache the answer, and return the effective list.
 *
 * ⚠⚠ THREE FALLBACKS, IN ORDER, AND THE LAST IS "EVERYTHING": what the server just said, then the
 * cached last-known-good, then the full catalogue. A till that loses its Reports tab because the
 * network blinked is worse than one showing a report an owner meant to hide.
 *
 * ⚠ Never throws, and never returns an empty list *because of a failure* — only because a shop
 * deliberately published nothing.
 */
export async function fetchPublished(
  get: (url: string) => Promise<unknown>,
  tillId?: string | null,
): Promise<string[]> {
  try {
    const url = tillId
      ? `/api/v1/reports/published?tillId=${encodeURIComponent(tillId)}`
      : "/api/v1/reports/published";

    const answer = (await get(url)) as { keys?: unknown } | null;
    const keys = answer?.keys;

    // ⚠ A missing/!array `keys` is "could not ask", NOT "nothing published" — do not cache it.
    if (!Array.isArray(keys)) return publishedOr(cachedPublished());

    const effective = publishedOr(keys.filter((k) => typeof k === "string"));
    cachePublished(effective);
    return effective;
  } catch {
    return publishedOr(cachedPublished());
  }
}

File: Plutus/Frontend/Plutus.Frontend.AppClient/Services/Reporting/PublishedReports.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace Plutus.Frontend.AppClient.Services.Reporting
{
    /// <summary>
    /// Which reports the portal has published to THIS till — ruling 5b(a), 2026-08-19.
    ///
    /// ⚠⚠ MATT: *"Portal shows which reports a till can show."* `ReportCatalogue` is now the SUPERSET;
    /// this decides which of it the operator is offered.
    ///
    /// ⚠⚠ **THE PUBLISH DECIDES THE MENU, THE PERMISSION DECIDES THE DOOR.** This class knows nothing
    /// about permissions and must not learn: `ReportsViewModel` applies both, and a report that is
    /// published but unreadable is not listed at all — never listed-and-refused, because a greyed row
    /// leaks what other roles can see.
    ///
    /// ⚠⚠ **THREE FALLBACKS, IN ORDER, AND THE LAST ONE IS "EVERYTHING".** A till that loses its Reports
    /// tab because the network blinked is worse than one showing a report an owner meant to hide:
    ///
    ///   1. what the server said just now;
    ///   2. the last-known-good cached in <c>Preferences</c> — survives a restart and an offline day;
    ///   3. the full catalogue, when this till has never once had an answer.
    ///
    /// ⚠ Which is why <see cref="Keys"/> is never empty *because of a failure*. It IS empty when a shop
    /// has deliberately published nothing — the cache stores that as a real, empty list, and the two
    /// states are distinguished by whether the key exists at all, exactly as the server distinguishes
    /// "no row" from "an empty row".
    /// </summary>
    internal static class PublishedReports
    {
        /// <summary>⚠ Versioned so a future change of shape cannot be read as a valid old value.</summary>
        private const string CacheKey = "plutus.reports.published.v1";

        private static readonly SemaphoreSlim Gate = new(1, 1);

        /// <summary>
        /// The reports this till may offer, in catalogue order.
        ///
        /// ⚠ Reads the CACHE only — never the network. It is called from viewmodel constructors and from
        /// the UI thread, and a screen must never block on a server to draw a menu.
        /// </summary>
        internal static IReadOnlyList<string> Keys
        {
            get
            {
                var cached = Load();

                // ⚠ NULL = never had an answer → every report. Not "none": see the class remarks.
                return SharedKernel.ReportCatalogue.PublishedOr(cached);
            }
        }

        /// <summary>
        /// Ask the server and update the cache. ⚠ Returns true when the answer CHANGED, so a screen knows
        /// whether it needs to rebuild its menu rather than rebuilding it on every tick.
        ///
        /// ⚠ Never throws. It runs on the settings cadence and from `OnAppearing`; an exception on either
        /// would be an `async void` escape, which is how this app has been killed before.
        /// </summary>
        internal static async Task<bool> RefreshAsync(CancellationToken ct = default)
        {
            // ⚠⚠ A DEADLINE, BECAUSE THIS GATE IS HELD ACROSS A NETWORK CALL. The body below
            // awaits the API, so a request that never returns holds this semaphore for ever and
            // every later refresh queues behind it in silence — the published-report list simply stops
            // updating, with no error anywhere. That is the `TillStoreAccess` fault in a second
            // place, and the convention it established is a deadline on every gate.
            //
            // ⚠ 30s, matching `TillStoreAccess.UseAsync`. One number, so nobody has to remember
            // which gate waits how long.
            // ⚠ AND IT RETURNS FALSE RATHER THAN THROWING — unlike the store gate, and
            // deliberately: this method runs on the settings cadence and its contract is NEVER
            // THROWS. "No change" is already how it reports being offline, and a timeout is the
            // same answer for the caller.
            if (!await Gate.WaitAsync(TimeSpan.FromSeconds(30), ct).ConfigureAwait(false))
            {
                Analytics.CrashLog.Write("PublishedReports.RefreshAsync(timeout)", new TimeoutException(
                    "The refresh gate did not become free within 30 seconds; this pass was skipped."));
                return false;
            }
            try
            {
                var api = await Connectivity.PlutusApi.GetAsync(ct).ConfigureAwait(false);
                if (api is null) return false;   // ⚠ Not connected is not a change.

                var credentials = await Connectivity.SecureDeviceCredentialStore.LoadAsync().ConfigureAwait(false);

                var answer = await api.GetPublishedReportsAsync(credentials?.TillId, ct).ConfigureAwait(false);

                // ⚠⚠ NULL IS "COULD NOT ASK", NOT "NOTHING PUBLISHED". Caching an empty list here would
                // empty the Reports tab on the first failed request and keep it empty offline.
                if (answer?.Keys is null) return false;

                var incoming = SharedKernel.ReportCatalogue.PublishedOr(answer.Keys);
                var current = Load();

                var changed = current is null || !current.SequenceEqual(incoming, StringComparer.Ordinal);
                if (changed) Save(incoming);

                return changed;
            }
            catch (Exception ex)
            {
                Analytics.CrashLog.Write("PublishedReports.Refresh", ex);
                return false;
            }
            finally
            {
                Gate.Release();
            }
        }

        /// <summary>
        /// The cached list, or <see langword="null"/> when this till has never had an answer.
        ///
        /// ⚠ A stored EMPTY list is returned as empty, not as null — "the owner published nothing" is a
        /// real state and must survive a restart.
        /// </summary>
        private static IReadOnlyList<string> Load()
        {
            try
            {
                var json = Microsoft.Maui.Storage.Preferences.Get(CacheKey, null);
                if (string.IsNullOrWhiteSpace(json)) return null;

                return JsonSerializer.Deserialize<List<string>>(json);
            }
            catch (Exception ex)
            {
                // ⚠ A corrupt cache reads as "never asked", which shows every report. Failing towards
                // MORE reports is the right direction for a menu; failing towards none is a broken till.
                Analytics.CrashLog.Write("PublishedReports.Load", ex);
                return null;
            }
        }

        private static void Save(IReadOnlyList<string> keys)
        {
            try
            {
                Microsoft.Maui.Storage.Preferences.Set(CacheKey, JsonSerializer.Serialize(keys));
            }
            catch (Exception ex)
            {
                Analytics.CrashLog.Write("PublishedReports.Save", ex);
            }
        }
    }
}
